enarx-keep-sev-shim/src/main.rs

```diff
@@ -1,47 +1,95 @@
 // SPDX-License-Identifier: Apache-2.0
 
-//! enarx-keep-sev-shim
+//! The SEV shim
 //!
-//! document
+//! This crate contains the system/kernel that handles the syscalls (and cpuid instructions)
+//! from the enclave code and might proxy them to the host.
 
 #![no_std]
 #![deny(clippy::all)]
+#![deny(clippy::integer_arithmetic)]
 #![deny(missing_docs)]
 #![cfg_attr(not(test), no_main)]
 
 #[cfg(test)]
 fn main() {}
 
-mod asm;
-mod no_std;
+pub mod addr;
+pub mod asm;
+pub mod hostcall;
+pub mod no_std;
+pub mod print;
+
+use addr::ShimVirtAddr;
+use hostcall::HostCall;
+
+use enarx_keep_sev_shim::BootInfo;
+use memory::Address;
+
+use core::convert::TryFrom;
 
 /// Defines the entry point function.
 ///
-/// The function must have the signature `fn(&'static ())) -> !`.
+/// The function must have the signature `fn(*mut BootInfo) -> !`.
 ///
 /// This macro just creates a function named `_start_main`, which the assembler
 /// stub will use as the entry point. The advantage of using this macro instead
 /// of providing an own `_start_main` function is that the macro ensures that the
 /// function and argument types are correct.
 macro_rules! entry_point {
     ($path:path) => {
-        #[allow(missing_docs)]
+        #[doc(hidden)]
         #[export_name = "_start_main"]
-        pub extern "C" fn __impl_start(boot_info: &'static ()) -> ! {
+        pub extern "C" fn __impl_start(boot_info: *mut BootInfo) -> ! {
             // validate the signature of the program entry point
-            let f: fn(&'static ()) -> ! = $path;
+            let f: fn(*mut BootInfo) -> ! = $path;
             f(boot_info)
         }
     };
 }
 
-entry_point!(kernel_main);
+entry_point!(shim_main);
+
+/// The entry point for the shim
+pub fn shim_main(boot_info: *mut BootInfo) -> ! {
+    HostCall::init(ShimVirtAddr::try_from(Address::from(boot_info).try_cast().unwrap()).unwrap());
+    eprintln!("Hello World!");
+    hostcall::shim_exit(0);
+}
+
+/// Halt Loop
+#[inline(always)]
+pub fn hlt_loop() -> ! {
+    loop {
+        x86_64::instructions::hlt();
+    }
+}
+
```

Comment:
The `hlt_loop()` function is a vestige of an external assembly function. I don't think it adds much value. Let's just use `loop { hlt() }` everywhere instead.

Reply:
ok